### PR TITLE
Named collections: finish replacing old code for storages

### PR DESCRIPTION
**Pull request by kssenii**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Storages/NamedCollectionsHelpers.cpp [hunk 1]
```diff
@@ -40,50 +42,87 @@ namespace
...
+                literal_value = evaluateConstantExpressionOrIdentifierAsLiteral(function_args[1], context);
+        }
+        catch (...)
+        {
+            if (fallback_to_ast_value)
+                return std::pair{key, function_args[1]};
+            throw;
+        }
 
+        auto value = literal_value->as<ASTLiteral>()->value;
```
Comment (antonio2368) [inline src/Storages/NamedCollectionsHelpers.cpp:63]: maybe have
```
return std::pair{key, Field(value)};
```

to be clear which variant type is used 🤔

File: src/Storages/NamedCollectionsHelpers.cpp [hunk 2]
```diff
@@ -40,50 +42,87 @@ namespace
...
-
     auto collection_copy = collection->duplicate();
 
+    if (asts.size() == 1)
+        return collection_copy;
+
     for (auto * it = std::next(asts.begin()); it != asts.end(); ++it)
     {
-        auto value_override = getKeyValueFromAST(*it);
+        auto value_override = getKeyValueFromAST(*it, complex_args != nullptr, context);
```
Comment (antonio2368) [inline src/Storages/NamedCollectionsHelpers.cpp:88]:
```suggestion
        auto value_override = getKeyValueFromAST(*it, /*fallback_to_ast_value*/ complex_args != nullptr, context);
```

File: src/Storages/NamedCollectionsHelpers.h [hunk 3]
```diff
@@ -16,16 +16,82 @@ namespace ErrorCodes
...
+MutableNamedCollectionPtr tryGetNamedCollectionWithOverrides(const Poco::Util::AbstractConfiguration & config, const std::string & config_prefix);
 
 HTTPHeaderEntries getHeadersFromNamedCollection(const NamedCollection & collection);
 
-template <typename RequiredKeys = std::unordered_set<std::string>,
-          typename OptionalKeys = std::unordered_set<std::string>>
+struct ExternalDatabaseEqualKeysSet
+{
+    static constexpr std::array<std::pair<std::string_view, std::string_view>, 5> equal_keys{
+        std::pair{"username", "user"}, std::pair{"database", "db"}, std::pair{"hostname", "host"}, std::pair{"addresses_expr", "host"}, std::pair{"addresses_expr", "hostname"}};
```
Comment (antonio2368) [inline src/Storages/NamedCollectionsHelpers.h:32]: you don't have to write `std::pair` inside array

File: src/Storages/NamedCollectionsHelpers.h [hunk 4]
```diff
@@ -16,16 +16,82 @@ namespace ErrorCodes
...
+};
+struct MongoDBEqualKeysSet
+{
+    static constexpr std::array<std::pair<std::string_view, std::string_view>, 4> equal_keys{
+        std::pair{"username", "user"}, std::pair{"database", "db"}, std::pair{"hostname", "host"}, std::pair{"table", "collection"}};
```
Comment (antonio2368) [inline src/Storages/NamedCollectionsHelpers.h:37]: same as above

File: src/Storages/StorageExternalDistributed.cpp [hunk 5]
```diff
@@ -237,102 +101,94 @@ void registerStorageExternalDistributed(StorageFactory & factory)
...
-        else if (engine_name == "PostgreSQL")
-            table_engine = StorageExternalDistributed::ExternalStorageEngine::PostgreSQL;
-        else
-            throw Exception(ErrorCodes::BAD_ARGUMENTS,
-                            "External storage engine {} is not supported for StorageExternalDistributed. "
-                            "Supported engines are: MySQL, PostgreSQL, URL",
-                            engine_name);
+        auto context = args.getLocalContext();
+        [[maybe_unused]] const auto & settings = context->getSettingsRef();
+        size_t max_addresses = context->getSettingsRef().glob_expansion_max_elements;
```
Comment (antonio2368) [inline src/Storages/StorageExternalDistributed.cpp:106]:
```suggestion
        const auto & settings = context->getSettingsRef();
        size_t max_addresses = settings.glob_expansion_max_elements;
```

or remove variable

File: src/Storages/NamedCollectionsHelpers.h (hunk 3 above)
Comment (kssenii) [reply src/Storages/NamedCollectionsHelpers.h:32]: you mean making it as `{"username", "user"}` instead of `std::pair{"username", "user"}`?

Comment (antonio2368) [reply src/Storages/NamedCollectionsHelpers.h:32]: yes, does it work?

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**